### PR TITLE
bump dkregistry-rs, futures, and all related dependencies; use async/await syntax for asynchronous code where possible

### PR DESCRIPTION
**Pull request by steveej**
TODO:
* [x] merge https://github.com/camallo/dkregistry-rs/pull/140 and bump revision here
* [x] self-review code
* [x] self-review git history

Closes #195.

### CODE REVIEW

File: cincinnati/src/plugins/mod.rs [hunk 1]
```diff
@@ -12,26 +12,25 @@ pub mod internal;
...
     pub use super::{build_plugins, deserialize_config, PluginSettings};
     pub use crate::{new_plugin, new_plugins};
     pub use futures_locks;
     pub use std::iter::FromIterator;
 }
 
-/// Convenience type to wrap other types in a Future
-pub type AsyncIO<T> = Box<dyn Future<Item = T, Error = Error> + Send>;
+/// Convenience type to wrap T
+pub type IO<T> = Fallible<T>;
```
Comment (steveej) [inline cincinnati/src/plugins/mod.rs]: @lucab WDYT about this? I think this one could be either renamed to `Result<T>` or simply dropped and all occurrences replaced by `Fallible`

Comment (lucab) [reply cincinnati/src/plugins/mod.rs]: :+1: on just dropping it.

Comment (steveej) [reply cincinnati/src/plugins/mod.rs]: It's gone!

File: graph-builder/Cargo.toml [hunk 2]
```diff
@@ -6,31 +6,31 @@ edition = "2018"
...
 [dependencies]
-actix = "^0.8.3"
-actix-web = "^1.0.2"
+actix = "^0.9"
+actix-web = "^2.0"
 chrono = "^0.4.7"
 cincinnati = { path = "../cincinnati" }
 commons = { path = "../commons" }
-dkregistry = { git = "https://github.com/camallo/dkregistry-rs.git", rev = "838947c22c3d434ab3176e9b4bae127e8ea670ce" }
+dkregistry = { git = "https://github.com/steveeJ-forks/dkregistry-rs.git", rev = "20c6c65ac4ae3455d41852c5ed6bfe3e5559cd44" }
```
Comment (steveej) [inline graph-builder/Cargo.toml]: TODO: bump this to use the upstreamed revision once it's merged

File: graph-builder/src/registry.rs [hunk 3]
```diff
@@ -352,110 +328,99 @@ fn cache_release<S: ::std::hash::BuildHasher>(
...
 
 // Get a stream of tags
-fn get_tags(
-    repo: String,
-    authenticated_client: &dkregistry::v2::Client,
-) -> impl Stream<Item = String, Error = Error> {
+async fn get_tags<'a, 'b: 'a>(
+    repo: &'b str,
+    authenticated_client: &'b dkregistry::v2::Client,
+) -> impl Stream<Item = Result<String, Error>> + 'a {
```
Comment (steveej) [inline graph-builder/src/registry.rs]: Could use `Fallible` here

File: graph-builder/src/registry.rs [hunk 4]
```diff
@@ -352,110 +328,99 @@ fn cache_release<S: ::std::hash::BuildHasher>(
...
 }
 
-fn get_manifest_and_ref(
+async fn get_manifest_and_ref(
     tag: String,
     repo: String,
     authenticated_client: &dkregistry::v2::Client,
-) -> impl Future<Item = (String, dkregistry::v2::manifest::Manifest, String), Error = failure::Error>
-{
+) -> Result<(String, dkregistry::v2::manifest::Manifest, String), failure::Error> {
```
Comment (steveej) [inline graph-builder/src/registry.rs:362]: Could use `Fallible` here

File: graph-builder/src/registry.rs [hunk 5]
```diff
@@ -352,110 +328,99 @@ fn cache_release<S: ::std::hash::BuildHasher>(
...
 
-fn find_first_release(
+async fn find_first_release(
     layer_digests: Vec<String>,
     authenticated_client: dkregistry::v2::Client,
     registry_host: String,
     repo: String,
     repo_tag: String,
-) -> impl Future<Item = (String, Option<Release>), Error = Error> {
+) -> Result<(String, Option<Release>), Error> {
```
Comment (steveej) [inline graph-builder/src/registry.rs]: Could use Fallible here

File: graph-builder/src/status.rs [hunk 6]
```diff
@@ -2,17 +2,13 @@
...
 
 /// Expose liveness status.
 ///
 /// Status:
 ///  * Live (200 code): The upstream scrape loop thread is running
 ///  * Not Live (500 code): everything else.
-pub fn serve_liveness(
-    req: HttpRequest,
-) -> Box<dyn Future<Item = HttpResponse, Error = failure::Error>> {
+pub async fn serve_liveness(req: HttpRequest) -> Result<HttpResponse, failure::Error> {
```
Comment (steveej) [inline graph-builder/src/status.rs]: Could use `Fallible` here

File: graph-builder/src/status.rs [hunk 7]
```diff
@@ -22,17 +18,16 @@ pub fn serve_liveness(
...
 
 /// Expose readiness status.
 ///
 /// Status:
 ///  * Ready (200 code): a JSON graph as the result of a successful scrape is available.
 ///  * Not Ready (500 code): no JSON graph available yet.
-pub fn serve_readiness(
-    req: HttpRequest,
-) -> Box<dyn Future<Item = HttpResponse, Error = failure::Error>> {
+pub async fn serve_readiness(req: HttpRequest) -> Result<HttpResponse, failure::Error> {
```
Comment (steveej) [inline graph-builder/src/status.rs]: Could use `Fallible` here

File: quay/src/v1/manifest.rs [hunk 8]
```diff
@@ -59,29 +57,29 @@ impl Into<(String, String)> for Label {
...
 impl Client {
     /// Fetch manifestref labels
-    pub fn get_labels<S: AsRef<str>>(
+    pub async fn get_labels<S: AsRef<str>>(
         &self,
         repository: S,
         manifest_ref: S,
         filter: Option<S>,
-    ) -> impl Future<Item = Vec<Label>, Error = Error> {
+    ) -> Result<Vec<Label>, Error> {
```
Comment (steveej) [inline quay/src/v1/manifest.rs:65]: Could use `Fallible` here

File: quay/src/v1/tag.rs [hunk 9]
```diff
@@ -30,25 +30,28 @@ pub struct Tag {
...
 impl Client {
     /// Fetch tags in a repository, in a streaming way.
-    pub fn stream_tags<S>(
-        &self,
+    pub async fn stream_tags<'a, 'b: 'a, S>(
+        &'b self,
         repository: S,
         only_active_tags: bool,
-    ) -> impl Stream<Item = Tag, Error = Error>
+    ) -> impl Stream<Item = Result<Tag, Error>> + 'a
```
Comment (steveej) [inline quay/src/v1/tag.rs]: Could use `Fallible` here

**Review by steveej — COMMENTED** [review]
* A couple styling/consistency and one unwrap finding
* The new async code still has some hard to read spots in the tests where double-boxes have to be used. I couldn't find a way to improve that, so I'm fine leaving it for now with a suggestion to revisit as the async ecosystem matures.

File: cincinnati/src/plugins/internal/arch_filter.rs [hunk 10]
```diff
@@ -82,85 +82,75 @@ lazy_static! {
...
-                                _ => true,
-                            }
-                        })
-                        .into_iter()
-                        .map(|(release_id, version)| {
-                            trace!("queuing '{}' for removal", version);
-                            release_id
-                        })
-                        .collect()
+        .map_err(failure::Error::from)?;
```
Comment (lucab) [inline cincinnati/src/plugins/internal/arch_filter.rs]: Is this error-mapping still required?

File: cincinnati/src/plugins/internal/cincinnati_graph_fetch.rs [hunk 11]
```diff
@@ -88,65 +89,63 @@ impl CincinnatiGraphFetchPlugin {
...
+            .send()
+            .map_err(|e| GraphError::FailedUpstreamFetch(e.to_string()))
+            .await?;
+
+        if !res.status().is_success() {
+            return Err(GraphError::FailedUpstreamFetch(res.status().to_string()).into());
+        }
+
+        let body = res
+            // .into_body()
```
Comment (lucab) [inline cincinnati/src/plugins/internal/cincinnati_graph_fetch.rs]: Leftover?

File: cincinnati/src/plugins/internal/metadata_fetch_quay.rs [hunk 12]
```diff
@@ -120,69 +117,64 @@ impl InternalPlugin for QuayMetadataFetchPlugin {
...
-                        };
-                    }
-                }
+            let quay_labels = client
+                .get_labels(
+                    repo.clone(),
+                    manifestref.clone(),
+                    Some(label_filter.clone()),
+                )
+                .await
```
Comment (lucab) [inline cincinnati/src/plugins/internal/metadata_fetch_quay.rs]: Can we try here? The logic below probably gets a bit easier to read outside of a `map()`.

File: commons/Cargo.toml [hunk 13]
```diff
@@ -5,19 +5,19 @@ authors = ["Stefan Junker <mail@stefanjunker.de>"]
...
 serde_json = "^1.0.34"
-tokio = "^0.1"
+tokio = "^0.2"
 url = "^1.7.2"
-futures = "^0.1"
+futures = "^0.3"
 
 [dev-dependencies]
-tokio = "^0.1"
+tokio = "^0.2"
```
Comment (lucab) [inline commons/Cargo.toml]: I think this can be dropped as it is already in `dependencies`, maybe?

File: graph-builder/src/graph.rs [hunk 14]
```diff
@@ -169,7 +164,7 @@ impl HasRegistry for State {
 }
 
 #[allow(clippy::useless_let_if_seq)]
-pub fn run<'a>(settings: &'a config::AppSettings, state: &State) -> ! {
+pub async fn run<'a>(settings: &'a config::AppSettings, state: &State) -> ! {
```
Comment (lucab) [inline graph-builder/src/graph.rs:167]: Whoa!

File: policy-engine/src/graph.rs [hunk 15]
```diff
@@ -35,93 +34,77 @@ pub(crate) fn register_metrics(registry: &Registry) -> Fallible<()> {
...
+        cincinnati::plugins::PluginIO::InternalIO(cincinnati::plugins::InternalIO {
+            graph: Default::default(),
+            parameters: plugin_params,
+        }),
+    )
+    .await
+    .map_err(|e| match e.downcast::<GraphError>() {
+        Ok(graph_error) => graph_error,
+        Err(other_error) => GraphError::FailedPluginExecution(other_error.to_string()),
+    })
```
Comment (lucab) [inline policy-engine/src/graph.rs]: I'd suggest breaking the chain here and moving the rest of the logic below out of combinators.

File: cincinnati/src/plugins/internal/arch_filter.rs (hunk 10 above)
Comment (steveej) [reply cincinnati/src/plugins/internal/arch_filter.rs]: Indeed not!

File: policy-engine/src/graph.rs (hunk 15 above)
Comment (steveej) [reply policy-engine/src/graph.rs]: That will increase the effort of logging the error from anywhere in the chain, as we do it now at the end of the chain. Unless you have a suggestion how to retain that mechanism somehow?

File: graph-builder/src/graph.rs (hunk 14 above)
Comment (steveej) [reply graph-builder/src/graph.rs:167]: It doesn't magically give us concurrency of what's going on inside though, we'll still have to work for that :wink:

File: policy-engine/src/graph.rs (hunk 15 above)
Comment (lucab) [reply policy-engine/src/graph.rs]: I see. Then fell free to keep it. An alternative could be to put the processing logic in its own function, and the error-logging logic as a wrapper one level above it.

Comment (steveej) [reply policy-engine/src/graph.rs]: I implemented your suggestion in a fixup commit. PTAL and let me know if you like that better in terms of readability. If so, I'm fine squashing it.

Comment (lucab) [reply policy-engine/src/graph.rs]: That looks good to me, but I know it's subjective taste. Leaving the final call to you.

Comment (steveej) [reply policy-engine/src/graph.rs]: I squashed the commit into the rest.

**Comment by steveej** [comment]
> ci/prow/rustfmt-1.37.0 	[dd421fb](https://github.com/openshift/cincinnati/commit/dd421fb1b217d7d942268bc036f890ea8c2a6801) 	[link](https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_cincinnati/195/pull-ci-openshift-cincinnati-master-rustfmt-1.37.0/108) 	`/test rustfmt-1.37.0`
> ci/prow/e2e 	[dd421fb](https://github.com/openshift/cincinnati/commit/dd421fb1b217d7d942268bc036f890ea8c2a6801) 	[link](https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_cincinnati/195/pull-ci-openshift-cincinnati-master-e2e/16) 	`/test e2e`

These failures are both expected at this point. The container image doesn't have the rust 1.37.0 toolchain anymore, and the e2e tests are still broken. @vrutkovs can we just skip these for this PR?

**Comment by vrutkovs** [comment]
Sure

/skip

**Comment by vrutkovs** [comment]
Ran a test version of this in a temp cluster and it fails with (when liveness/readyness probes are called):

<details>
<summary>graph-builder</summary>

```
[2020-01-31T13:40:18Z DEBUG graph_builder::graph] graph update triggered
[2020-01-31T13:40:20Z TRACE graph_builder::registry] processing: redhat/openshift-cincinnati-test-public-manual:0.0.0
[2020-01-31T13:40:20Z TRACE graph_builder::registry] processing: redhat/openshift-cincinnati-test-public-manual:0.0.1
[2020-01-31T13:40:20Z TRACE graph_builder::registry] Downloading layer sha256:28929abdc540be2da178b0037ec005f667dedb93d6ea7c6de990e50184cf8966...
[2020-01-31T13:40:21Z TRACE graph_builder::registry] 0.0.0: Looking for release-manifests/release-metadata in archive sha256:28929abdc540be2da178b0037ec005f667dedb93d6ea7c6de990e50184cf8966 with 203 bytes
[2020-01-31T13:40:21Z TRACE graph_builder::registry] Caching release metadata for new tag 0.0.0
[2020-01-31T13:40:21Z TRACE graph_builder::registry] Downloading layer sha256:12bf6b4e26018450308527e986c6bdc2403e975a5f87166f3b175c287167414a...
[2020-01-31T13:40:22Z TRACE graph_builder::registry] 0.0.1: Looking for release-manifests/release-metadata in archive sha256:12bf6b4e26018450308527e986c6bdc2403e975a5f87166f3b175c287167414a with 238 bytes
[2020-01-31T13:40:22Z TRACE graph_builder::registry] Caching release metadata for new tag 0.0.1
[2020-01-31T13:40:22Z TRACE graph_builder::graph] Adding a release to the graph 'Release { source: "quay.io/redhat/openshift-cincinnati-test-public-manual@sha256:a264db3ac5288c9903dc3db269fca03a0b122fe4af80b57fc5087b329995013d", metadata: Metadata { kind: V0, version: Version { major: 0, minor: 0, patch: 0, pre: [], build: [AlphaNumeric("amd64")] }, previous: [], next: [], metadata: {"io.openshift.upgrades.graph.release.arch": "amd64", "io.openshift.upgrades.graph.release.manifestref": "sha256:a264db3ac5288c9903dc3db269fca03a0b122fe4af80b57fc5087b329995013d"} } }'
[2020-01-31T13:40:22Z TRACE graph_builder::graph] Adding a release to the graph 'Release { source: "quay.io/redhat/openshift-cincinnati-test-public-manual@sha256:73df5efa869eaf57d4125f7655e05e1a72b59d05e55fea06d3701ea5b59234ff", metadata: Metadata { kind: V0, version: Version { major: 0, minor: 0, patch: 1, pre: [], build: [AlphaNumeric("amd64")] }, previous: [Version { major: 0, minor: 0, patch: 0, pre: [], build: [] }], next: [], metadata: {"kind": "test", "io.openshift.upgrades.graph.release.arch": "amd64", "io.openshift.upgrades.graph.release.manifestref": "sha256:73df5efa869eaf57d4125f7655e05e1a72b59d05e55fea06d3701ea5b59234ff"} } }'
[2020-01-31T13:40:22Z DEBUG graph_builder::graph] graph update completed, 2 valid releases
thread 'actix-rt:worker:0' panicked at 'the request has no app_data attached. this is a bug.', src/libcore/option.rs:1185:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
thread 'actix-rt:worker:1' panicked at 'the request has no app_data attached. this is a bug.', src/libcore/option.rs:1185:5
```
</details>

Perhaps it was a poor rebase on master though, so I'd rather wait for e2e tests to land in master

**Comment by steveej** [comment]
> Ran a test version of this in a temp cluster and it fails with:

That's interesting, thanks for testing this. I tried to cover this in a test but I must've missed something. Please request changes on the PR to block it.

**Review by vrutkovs — CHANGES_REQUESTED** [review]
/hold

Seems to crash on liveness/readyness checks in real environment

**Comment by steveej** [comment]
> Ran a test version of this in a temp cluster and it fails with:

I can't reproduce this error by running this in three different shells on 80f2b6c4199c0e5a1c63861a07a76f9b2523ee4d:

graph-builder:
```console
$ target/debug/graph-builder --upstream.registry.pause_secs 30 --address 0.0.0.0 --registry 'https://quay.io' --repository redhat/openshift-cincinnati-test-public-multiarch-manual -vvv --service.path_prefix api/upgrades_info/ --credentials-file /home/steveej/.docker/config.json
```

policy-engine:
```console
$ target/debug/policy-engine -vvvv --service.address 0.0.0.0 --service.path_prefix api/upgrades_info/ --upstream.cincinnati.url 'http://localhost:8080/api/upgrades_info/v1/graph' --service.mandatory_client_parameters=channel
```

<details>
<summary>
curl:

```console
curl --header 'Accept:application/json' http://localhost:8081/api/upgrades_info/v1/graph?channel='test'\&arch='amd64' | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1950  100  1950    0     0  81250      0 --:--:-- --:--:-- --:--:-- 81250
```

</summary>

```json
{
  "nodes": [
    {
      "version": "0.0.0",
      "payload": "quay.io/redhat/openshift-cincinnati-test-public-multiarch-manual@sha256:6f1a8384ea813d6d3c1c609676b2ef3535d2bdf6489e29279d104bdc7e8a8ef7",
      "metadata": {
        "io.openshift.upgrades.graph.release.channels": "test",
        "io.openshift.upgrades.graph.release.manifestref": "sha256:6f1a8384ea813d6d3c1c609676b2ef3535d2bdf6489e29279d104bdc7e8a8ef7"
      }
    },
    {
      "version": "0.0.1",
      "payload": "quay.io/redhat/openshift-cincinnati-test-public-multiarch-manual@sha256:f03bc60a382c5b24520c3dd595f3e2e8210cb555e585a11349c68d75314a0bd5",
      "metadata": {
        "io.openshift.upgrades.graph.release.channels": "test",
        "io.openshift.upgrades.graph.release.manifestref": "sha256:f03bc60a382c5b24520c3dd595f3e2e8210cb555e585a11349c68d75314a0bd5"
      }
    },
    {
      "version": "0.0.2",
      "payload": "quay.io/redhat/openshift-cincinnati-test-public-multiarch-manual@sha256:964a52554097912533f0d3b06c54e4a13b6e054716c0b44339104708b31660dc",
      "metadata": {
        "io.openshift.upgrades.graph.release.channels": "test",
        "io.openshift.upgrades.graph.release.manifestref": "sha256:964a52554097912533f0d3b06c54e4a13b6e054716c0b44339104708b31660dc"
      }
    },
    {
      "version": "0.0.4",
      "payload": "quay.io/redhat/openshift-cincinnati-test-public-multiarch-manual@sha256:062db78b55bff758bc1bc175e06c7526f6d646f6a0656acd1b23e2ee02282699",
      "metadata": {
        "io.openshift.upgrades.graph.release.channels": "test",
        "io.openshift.upgrades.graph.release.manifestref": "sha256:062db78b55bff758bc1bc175e06c7526f6d646f6a0656acd1b23e2ee02282699"
      }
    },
    {
      "version": "0.0.3",
      "payload": "quay.io/redhat/openshift-cincinnati-test-public-multiarch-manual@sha256:f777d12975d4337b0f081c2c6404915b6074d6526bdb151928e4acd47dc0fc7d",
      "metadata": {
        "io.openshift.upgrades.graph.previous.add": "0.0.1",
        "io.openshift.upgrades.graph.previous.remove": "0.0.2",
        "io.openshift.upgrades.graph.release.manifestref": "sha256:f777d12975d4337b0f081c2c6404915b6074d6526bdb151928e4acd47dc0fc7d",
        "io.openshift.upgrades.graph.release.channels": "test"
      }
    }
  ],
  "edges": [
    [
      0,
      1
    ],
    [
      1,
      2
    ],
    [
      1,
      4
    ],
    [
      4,
      3
    ]
  ]
}
```
</details>

The result seems fine to me.

**Comment by vrutkovs** [comment]
Happens when liveness endpoint is called:
```
...
  10: core::panicking::panic_fmt
  11: core::option::expect_failed
  12: core::option::Option<T>::expect
             at /builddir/build/BUILD/rustc-1.40.0-src/src/libcore/option.rs:345
  13: graph_builder::status::serve_liveness::{{closure}}
             at ./graph-builder/src/status.rs:13
...
```

**Comment by steveej** [comment]
> Happens when liveness endpoint is called:

Fixed in ba7fe2bff2a3a24ff395f670ee99510ceac1bfaf.

**Comment by wking** [comment]
Will need a rebase to resolve `Cargo.lock` conflicts (in the wake of #202?).

**Review by vrutkovs — APPROVED** [review]
/lgtm

**Pull request merged**